Jenkinsfile

```diff
@@ -0,0 +1,164 @@
+def setupCC(failOnError = true) {
+    errorStr = failOnError ? "-Werror " : ""
+    "echo CC=${env.CXX} ${errorStr}> make/local"
+}
+
+def setup(String pr, Boolean failOnError = true) {
+    errorStr = failOnError ? "-Werror " : ""
+    script = """
+        make math-revert
+        make clean-all
+        git clean -xffd
+        ${setupCC(failOnError)}
+    """
+    if (pr != '')  {
+        prNumber = pr.tokenize('-').last()
+        script += """ 
+            cd lib/stan_math
+            git fetch https://github.com/stan-dev/math +refs/pull/${prNumber}/merge:refs/remotes/origin/pr/${prNumber}/merge
+            git checkout refs/remotes/origin/pr/${prNumber}/merge
+        """
+    }
+    return script
+}
+
+def mailBuildResults(String label, additionalEmails='') {
+    emailext (
+        subject: "[StanJenkins] ${label}: Job '${env.JOB_NAME} [${env.BUILD_NUMBER}]'",
+        body: """${label}: Job '${env.JOB_NAME} [${env.BUILD_NUMBER}]': Check console output at ${env.BUILD_URL}""",
+        recipientProviders: [[$class: 'RequesterRecipientProvider']],
+        to: "${env.CHANGE_AUTHOR_EMAIL}, ${additionalEmails}"
+    )
+}
+
+def runTests(String testPath) {
+    "runTests.py -j${env.PARALLEL} ${testPath} || echo ${testPath} failed"
+}
+
+def updateUpstream(String upstreamRepo) {
+    if (env.BRANCH_NAME == 'develop') {
+        sh "curl -O https://raw.githubusercontent.com/stan-dev/ci-scripts/master/jenkins/create-${upstreamRepo}-pull-request.sh"
+        sh "sh create-${upstreamRepo}-pull-request.sh"
+    }
+}
+
+pipeline {
+    agent none
+    parameters {
+        string(defaultValue: '', name: 'math_pr', description: "Leave blank "
```

Comment:
Is the idea to provide the changed math branch to this Stan PR?

Mind describing how to deal with a math branch that breaks Stan or a Stan branch that breaks CmdStan?

Reply:
Yeah, so that if you wanted to run the Stan suite (likely on its own PR or branch) against a specific math PR, you could do that here in a way that's pretty similar to how we currently do it ("Build with parameters" and, now, enter just the github PR number). CmdStan has two similar parameters, `stan_pr` and `math_pr` that you enter in a similar way on the build with parameters screen:
![image](https://user-images.githubusercontent.com/805356/31665627-77596930-b317-11e7-8ab2-d8570a769870.png)


I need to add a doc string to those parameters in CmdStan, woops. 

So this pipelines thing does not immediately seem to be able to improve the situation regarding these multi-repo breaking changesets :(

Reply:
Stan and Math could also take parameters indicating which CmdStan (and Stan) PRs to build against when running the upstream tests, which is maybe slightly better because then the test results show up against the lower repo's pull request (on github) as that check or X and it runs them all at once. But there's no way to automatically do that, either, and these parameters are necessary because these jobs are triggered by the tests downstream with the parameters filled in - this is new. Currently we have a sort of more basic single-job "upstream" test defined (e.g. Upstream Stan in Math: http://d1m1s1b1.stat.columbia.edu:8080/job/Math%20Pull%20Request%20-%20Upstream%20-%20Stan/) but what I have happening here is triggering a complete build of the upstream pipeline (for now under this "downstream tests" "job" shown on this page here: http://d1m1s1b1.stat.columbia.edu:8080/job/CmdStan/)

Reply:
Great! Thanks for the clarificaion